src/c/src/node.c

```diff
@@ -0,0 +1,373 @@
+    nodeset_add(other->edges, self);
+  }
+}
+
+void add_child(Node *self, Node *child) {
+  if (!self || !child)
+    return;
+
+  child->parent = self;
+  nodeset_add(self->children, child);
+}
+
+bool is_root(Node *node) { return node ? node->parent == NULL : false; }
+
+bool is_leaf(Node *node) {
+  return node ? nodeset_size(node->children) == 0 : true;
+}
+
+int height(Node *node) {
+  if (!node || is_leaf(node))
+    return 0;
+  int max = 0;
+  for (size_t i = 0; i < nodeset_size(node->children); i++) {
+    int h = height(node->children->nodes[i]);
+    if (h > max)
+      max = h;
+  }
+  return max + 1;
+}
+
+int depth(Node *node) {
+  if (!node)
+    return -1;
+  int d = 0;
+  while (node->parent) {
+    d++;
+    node = node->parent;
+  }
+  return d;
+}
+
+int num_leaves(Node *node) {
+  if (!node)
+    return 0;
+  if (is_leaf(node))
+    return 1;
+  int count = 0;
+  for (size_t i = 0; i < nodeset_size(node->children); i++) {
+    count += num_leaves(node->children->nodes[i]);
+  }
+  return count;
+}
+
+int num_nodes(Node *node) {
+  if (!node)
+    return 0;
+  int count = 1;
+  for (size_t i = 0; i < nodeset_size(node->children); i++) {
+    count += num_nodes(node->children->nodes[i]);
+  }
+  return count;
+}
+
+int diameter(Node *node) {
+  if (!node)
+    return 0;
+  int max1 = 0, max2 = 0;
+  for (size_t i = 0; i < nodeset_size(node->children); i++) {
+    int h = height(node->children->nodes[i]);
+    if (h > max1) {
+      max2 = max1;
+      max1 = h;
+    } else if (h > max2) {
+      max2 = h;
+    }
+  }
+  int max_child_dia = 0;
+  for (size_t i = 0; i < nodeset_size(node->children); i++) {
+    int d = diameter(node->children->nodes[i]);
+    if (d > max_child_dia)
+      max_child_dia = d;
+  }
+  return max1 + max2 + 1 > max_child_dia ? max1 + max2 + 1 : max_child_dia;
+}
+
+// BST Operations
+Node *bst_insert(Node *root, void *value, double node_id,
+                 int (*compare)(const void *a, const void *b)) {
+  if (!root) {
+    return node_create(value, node_id);
+  }
+
+  int cmp = compare(value, root->value);
+  if (cmp < 0) {
+    root->left = bst_insert(root->left, value, node_id, compare);
+  } else if (cmp > 0) {
+    root->right = bst_insert(root->right, value, node_id, compare);
+  }
+  // If equal, don't insert duplicate
+
+  return root;
+}
+
+Node *bst_search(Node *root, void *value,
+                 int (*compare)(const void *a, const void *b)) {
+  if (!root)
+    return NULL;
+
+  int cmp = compare(value, root->value);
+  if (cmp == 0)
+    return root;
+  else if (cmp < 0)
+    return bst_search(root->left, value, compare);
+  else
+    return bst_search(root->right, value, compare);
+}
+
+Node *bst_find_min(Node *root) {
+  if (!root)
+    return NULL;
+  while (root->left) {
+    root = root->left;
+  }
+  return root;
+}
+
+Node *bst_find_max(Node *root) {
+  if (!root)
+    return NULL;
+  while (root->right) {
+    root = root->right;
+  }
+  return root;
+}
```

Comment:
1. BST functions should not be in `node.c`
2. These are sub-optimal functions for finding max and min. let's drop them for now.